### ISSUE TITLE
Currently no way to scaffold Enums

### ISSUE THREAD

**Issue by weaversam8** [redwoodjs/redwood#313]
If you create an [Enum](https://github.com/prisma/prisma2/blob/master/docs/data-modeling.md#enums) in your database schema (within `schema.prisma`), and then scaffold out your main types (by using `yarn rw g scaffold TypeName`, for each), `yarn rw dev` will still fail, since GraphQL won't be able to find schemas defined for the enums your types are referencing.

## Example

In the following example:

```prisma
type Shoe {
  id String @id @default(cuid())
  name String
  size Int
  color Color
}

enum Color {
  RED
  GREEN
  BLUE
}
```

Running `yarn rw g scaffold Shoe` will execute successfully, but then running `yarn rw dev` will give you an error, with the message `Error: Can't find type Color`. 

- Running `yarn rw g scaffold Color` won't work, since `Color` isn't a full type, but an enum.
- Running `yarn rw g sdl Color` won't work, giving the error message: ``Cannot create property 'context' on string 'No schema definition found for `Color`'``.

## Workaround

A workaround for this is to add the enum types to your `.sdl` files directly. In the above example, you can simply add the following lines to your `shoes.sdl.js` file:

```gql
enum Color {
  RED
  GREEN
  BLUE
}
```

## Long term solution

Long term, I see two ways of dealing with this.

1. **Intelligent:** an "intelligent" solution might generate and include the SDL for the `Color` enum in `shoes.sdl.js` automatically. This might be challenging, since we would need to "know" that `Color` is an enum and referenced by the `Shoe` model.
    - Additionally, if multiple models use the `Color` enum, where should the SDL for that live? If it were duplicated across multiple files, would that cause an error when the SDL files are merged? If it wouldn't cause an error, would it even be a good idea?
2. **Naive:** the naive solution would be simply to extend the `sdl` command in the [`@redwoodjs/api`](https://github.com/redwoodjs/redwood/tree/master/packages/api) package to generate the sdl in a new file (perhaps called `color.sdl.js`). **This seems like the quickest solution that we could write a PR for right now.**
    - One open question with this: if we did generate such a file, is there anywhere else where it needs to be added? Is there a list of SDL files that Redwood "imports" when it merges the master schema, or does it crawl `api/src/graphql` and search for files that meet the glob `*.sdl.js`?

**Comment by thedavidprice** [comment]
@cannikin and @peterp please take a look at this when you get a chance.

Definitely leaning "Naive" myself per comments above. I don't have a confident answer to the question, "does this file need to be added anywhere else?", but I believe it's "no".

Also, there's overlap topically with #323 specifically regarding "To what extent will our generators support Prisma2 features?" And, regardless, "How do we make the generator(s) more intelligent to handle cases outside our scope?"

**Comment by weaversam8** [comment]
I now have a confident answer to the question "does this file need to be added anywhere else?"

[This line](https://github.com/redwoodjs/create-redwood-app/blob/572e3a477c0666d34eb970b9e809f6877b595b9c/api/src/functions/graphql.js#L8) in `api/src/functions/graphql.js` (generated by `create-redwood-app`) calls the [`importAll`](https://github.com/redwoodjs/redwood/blob/c71efc09fecc705dacd17d5bdb700afe3536da96/packages/api/src/importAll.macro.js) macro, which should pick up any file matching the glob `*.sdl.js`, I'm _pretty sure._

**Comment by Rosenberg96** [comment]
Chipping in here, since I experienced the lack of support for this in my side project.

@thedavidprice @weaversam8 I looked into this, and if you inspect the `DMMF` returned by the `Prisma` project you will see that this exposes the following definitions:

```
export interface Datamodel {
    models: Model[]
    enums: DatamodelEnum[]
}

export interface DatamodelEnum {
    name: string
    values: EnumValue[]
    dbName?: string | null
  }

```
We can use these definitions combined with a property on each `Model.type`, which in this case is `enum`. 

At the time of writing I have managed to actually generate the correct `Enum` type definitions in the relevant `sdl.js` file. 🚀 

However, to your follow-up point. I agree that if several scaffold operations uses the same `enum` definition, we will run into issue. I personally, find it most naturally to have the enum defined in the scheme definition it is used, but we need some way to handle that multiple definitions may use the same `enum`. I have not figured out a nice way to mitigate this.

Alternatively, we can ensure that the `importAllMacro.js` ensures uniqueness, but that approach may also come with some caveats, since scaffolding operations across model definitions may be out of sync.

There are a few things left to resolve in my PR - especially in regards to the `sdl.template.js`, but I think I will open a PR to show my progress so far.

Finally, we should also alter the `scaffold` generation in the `NameForm.js.template`. My suggestion is to render Enum types with a `select` or similar.

**Comment by weaversam8** [comment]
@Rosenberg96 - what if we generate the enum SDL to its own `*.sdl.js` file. Might be a little bit extra, but it'll avoid having to ask the question of where it belongs.

> scaffolding operations across model definitions may be out of sync
I'm not sure if we need to worry about this immediately, I think right now the scaffolding generator ignores any relations to other models right now. 

If I run `yarn rw g scaffold User` and the User type references Post, Post won't get scaffolded right now. (@cannikin please correct me if I'm wrong.)

Long term, obviously, we probably want to be able to generate all the linked SDL for a model automatically, but that's probably out of scope for this issue.

**Comment by Rosenberg96** [comment]
@weaversam8 While it is possible to output the `enum` to its own `*.sdl.js` file I found no easy way to do this, while outputting to the same definition file was relatively easy. Others may feel differently though.

To my earlier point: 

>   I agree that if several scaffold operations uses the same enum definition, we will run into issues.

I actually tested this across several scaffolded datamodels and while a change to the enum definition in the `schema.prisma`is only applied to the subsequent scaffolding operations of that data model definition, we still in the merged `schema` given to the GraphQL instance actually support the set of all available enum values.

This implies that if two data model definitions utilise the same `enum` definition - we expose the super-set of the values defined with the same enum name. I suspect that this is due to the `importAllMacro.js`. I found this quite neat.

I have a working version of the 'easy' approach. Maybe some one could advise which direction to pursue?
